File: libs/common/src/events/consumer.ts
/**
 * Event consumer for RabbitMQ
 */

import * as amqp from 'amqplib';

export type EventHandler = (eventData: Record<string, any>) => Promise<void>;

export interface EventConsumerConfig {
  url: string;
  queue: string;
  prefetchCount?: number;
}

export class EventConsumer {
  private connection?: amqp.Connection;
  private channel?: amqp.Channel;
  private readonly handlers: Map<string, EventHandler[]> = new Map();
  private running = false;

  constructor(
    private readonly config: EventConsumerConfig,
  ) {}

  /**
   * Connect to RabbitMQ
   */
  async connect(): Promise<void> {
    try {
      console.log('Connecting to RabbitMQ', this.config.queue);

      this.connection = (await amqp.connect(this.config.url)) as any;
      this.channel = (await (this.connection as any).createChannel()) as any;

      await (this.channel as any).prefetch(this.config.prefetchCount || 10);
      await (this.channel as any).assertQueue(this.config.queue, { durable: true });

      console.log('Connected to RabbitMQ', this.config.queue);
    } catch (error) {
      console.error('Failed to connect to RabbitMQ', error);
      throw error;
    }
  }

  /**
   * Disconnect from RabbitMQ
   */
  async disconnect(): Promise<void> {
    try {
      this.running = false;
      if (this.channel) {
        await (this.channel as any).close();
      }
      if (this.connection) {
        await (this.connection as any).close();
      }
      console.log('Disconnected from RabbitMQ');
    } catch (error) {
      console.error('Error disconnecting from RabbitMQ', error);
    }
  }

  /**
   * Register an event handler
   */
  registerHandler(eventType: string, handler: EventHandler): void {
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, []);
    }
    this.handlers.get(eventType)!.push(handler);
    console.log('Registered event handler', eventType);
  }

  /**
   * Process a message from the queue
   */
  private async processMessage(message: amqp.ConsumeMessage): Promise<void> {
    try {
      // Parse message body
      const eventData = JSON.parse(message.content.toString());
      const eventType = eventData.eventType;

      console.log('Processing event', eventType);

      // Get handlers for this event type
      const handlers = this.handlers.get(eventType) || [];
      if (handlers.length === 0) {
        console.warn('No handlers registered for event type', eventType);
        // Acknowledge message even if no handlers
        (this.channel as any)?.ack(message);
        return;
      }

      // Execute all handlers
      for (const handler of handlers) {
        try {
          await handler(eventData);
          console.log('Handler executed successfully', eventType);
        } catch (error) {
          console.error('Handler failed', error);
          // Reject message and requeue
          (this.channel as any)?.nack(message, false, true);
          return;
        }
      }

      // Acknowledge message after all handlers succeed
      (this.channel as any)?.ack(message);
    } catch (error) {
      console.error('Failed to process message', error);
      // Reject message without requeue for invalid messages
      (this.channel as any)?.nack(message, false, false);
    }
  }

  /**
   * Start consuming messages
   */
  async start(): Promise<void> {
    if (!this.channel) {
      throw new Error('Consumer not connected. Call connect() first.');
    }

    this.running = true;
    console.log('Starting event consumer', this.config.queue);

    try {
      await (this.channel as any).consume(
        this.config.queue,
        (message: amqp.ConsumeMessage | null) => {
          if (message) {
            this.processMessage(message).catch((error) => {
              console.error('Error processing message', error);
            });
          }
        },
        { noAck: false },
      );

      console.log('Event consumer started', this.config.queue);
    } catch (error) {
      console.error('Error in event consumer', error);
      throw error;
    }
  }

  /**
   * Stop consuming messages
   */
  async stop(): Promise<void> {
    this.running = false;
    console.log('Stopping event consumer');
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<{
    healthy: boolean;
    queue?: string;
    running?: boolean;
    handlers?: Record<string, number>;
    message?: string;
    error?: string;
  }> {
    try {
      if (!this.connection || !this.channel) {
        return {
          healthy: false,
          message: 'Not connected to RabbitMQ',
        };
      }

      const handlerCounts: Record<string, number> = {};
      this.handlers.forEach((handlers, eventType) => {
        handlerCounts[eventType] = handlers.length;
      });

      return {
        healthy: true,
        queue: this.config.queue,
        running: this.running,
        handlers: handlerCounts,
      };
    } catch (error) {
      return {
        healthy: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

File: libs/python/events/consumer.py
"""Event consumer for RabbitMQ."""

import asyncio
import json
from typing import Any, Callable, Dict, List, Optional, Protocol

import aio_pika
from aio_pika import Connection, Channel, Queue, IncomingMessage
from aio_pika.abc import AbstractRobustConnection

from common.logger import LoggerMixin


class EventHandler(Protocol):
    """Protocol for event handlers."""

    async def __call__(self, event_data: Dict[str, Any]) -> None:
        """Handle an event.

        Args:
            event_data: Event data as dictionary
        """
        ...


class EventConsumer(LoggerMixin):
    """Event consumer for consuming domain events from RabbitMQ."""

    def __init__(
        self,
        rabbitmq_url: str,
        queue_name: str,
        prefetch_count: int = 10,
    ):
        """Initialize event consumer.

        Args:
            rabbitmq_url: RabbitMQ connection URL
            queue_name: Name of the queue to consume from
            prefetch_count: Number of messages to prefetch
        """
        super().__init__()
        self.rabbitmq_url = rabbitmq_url
        self.queue_name = queue_name
        self.prefetch_count = prefetch_count
        self._connection: Optional[AbstractRobustConnection] = None
        self._channel: Optional[Channel] = None
        self._queue: Optional[Queue] = None
        self._handlers: Dict[str, List[EventHandler]] = {}
        self._running = False

    async def connect(self) -> None:
        """Connect to RabbitMQ."""
        try:
            self.logger.info(
                "Connecting to RabbitMQ",
                queue=self.queue_name,
            )
            self._connection = await aio_pika.connect_robust(self.rabbitmq_url)
            self._channel = await self._connection.channel()
            await self._channel.set_qos(prefetch_count=self.prefetch_count)
            self._queue = await self._channel.declare_queue(
                self.queue_name,
                durable=True,
            )
            self.logger.info(
                "Connected to RabbitMQ",
                queue=self.queue_name,
            )
        except Exception as e:
            self.logger.error(
                "Failed to connect to RabbitMQ",
                error=str(e),
                queue=self.queue_name,
            )
            raise

    async def disconnect(self) -> None:
        """Disconnect from RabbitMQ."""
        try:
            self._running = False
            if self._channel:
                await self._channel.close()
            if self._connection:
                await self._connection.close()
            self.logger.info("Disconnected from RabbitMQ")
        except Exception as e:
            self.logger.error(
                "Error disconnecting from RabbitMQ",
                error=str(e),
            )

    def register_handler(
        self,
        event_type: str,
        handler: EventHandler,
    ) -> None:
        """Register an event handler.

        Args:
            event_type: Type of event to handle
            handler: Handler function
        """
        if event_type not in self._handlers:
            self._handlers[event_type] = []
        self._handlers[event_type].append(handler)
        self.logger.info(
            "Registered event handler",
            event_type=event_type,
            handler=handler.__name__,
        )

    async def _process_message(self, message: IncomingMessage) -> None:
        """Process a message from the queue.

        Args:
            message: Incoming message
        """
        async with message.process():
            try:
                # Parse message body
                event_data = json.loads(message.body.decode())
                event_type = event_data.get("event_type")

                self.logger.info(
                    "Processing event",
                    event_id=event_data.get("event_id"),
                    event_type=event_type,
                    correlation_id=event_data.get("correlation_id"),
                )

                # Get handlers for this event type
                handlers = self._handlers.get(event_type, [])
                if not handlers:
                    self.logger.warning(
                        "No handlers registered for event type",
                        event_type=event_type,
                    )
                    return

                # Execute all handlers
                for handler in handlers:
                    try:
                        await handler(event_data)
                        self.logger.info(
                            "Handler executed successfully",
                            event_type=event_type,
                            handler=handler.__name__,
                        )
                    except Exception as e:
                        self.logger.error(
                            "Handler failed",
                            event_type=event_type,
                            handler=handler.__name__,
                            error=str(e),
                        )
                        # Re-raise to trigger message requeue
                        raise

            except json.JSONDecodeError as e:
                self.logger.error(
                    "Failed to parse message body",
                    error=str(e),
                )
                # Don't requeue invalid messages
            except Exception as e:
                self.logger.error(
                    "Failed to process message",
                    error=str(e),
                )
                raise

    async def start(self) -> None:
        """Start consuming messages."""
        if not self._queue:
            raise RuntimeError("Consumer not connected. Call connect() first.")

        self._running = True
        self.logger.info(
            "Starting event consumer",
            queue=self.queue_name,
        )

        try:
            await self._queue.consume(self._process_message)
            self.logger.info(
                "Event consumer started",
                queue=self.queue_name,
            )

            # Keep running until stopped
            while self._running:
                await asyncio.sleep(1)

        except Exception as e:
            self.logger.error(
                "Error in event consumer",
                error=str(e),
            )
            raise

    async def stop(self) -> None:
        """Stop consuming messages."""
        self._running = False
        self.logger.info("Stopping event consumer")

    async def health_check(self) -> Dict[str, Any]:
        """Check health of the consumer.

        Returns:
            Health check result
        """
        try:
            if not self._connection or self._connection.is_closed:
                return {
                    "healthy": False,
                    "message": "Not connected to RabbitMQ",
                }

            return {
                "healthy": True,
                "queue": self.queue_name,
                "running": self._running,
                "handlers": {
                    event_type: len(handlers)
                    for event_type, handlers in self._handlers.items()
                },
            }
        except Exception as e:
            return {
                "healthy": False,
                "error": str(e),
            }

